### PR TITLE
Add Type Hints to modeling_utils.py Closes #3911

### LINKED ISSUE #3911: Type Hints for modeling_utils.py
add [type hints](https://docs.python.org/3/library/typing.html) to the methods for better readability and autocomplete.

If possible, check your work with an automated tool and tell us how you did it!

### PR DESCRIPTION
**Pull request by bglearning**
Add Type Hints to methods in `modeling_utils.py`

Note: The coverage isn't 100%. Mostly skipped internal methods (and some I wasn't sure of).

### CODE REVIEW

**Review by LysandreJik — APPROVED** [review]
This is great, thanks!

**Comment by patrickvonplaten** [comment]
I think this is good for merge, no? @julien-c

File: src/transformers/modeling_utils.py [hunk 1]
```diff
@@ -17,7 +17,7 @@
 
 import logging
 import os
-from typing import Callable, Tuple
+from typing import Callable, Sequence, Tuple
```
Comment (julien-c) [inline src/transformers/modeling_utils.py:20]: we usually use `typing.Iterable`, any reason to use `Sequence` here instead?

File: src/transformers/modeling_utils.py [hunk 2]
```diff
@@ -267,7 +267,7 @@ def get_input_embeddings(self):
         else:
             raise NotImplementedError
 
-    def set_input_embeddings(self, value):
+    def set_input_embeddings(self, value: nn.Module):
```
Comment (julien-c) [inline src/transformers/modeling_utils.py:270]: is this not an `nn.Layer` or even a `nn.Embedding`?

File: src/transformers/modeling_utils.py [hunk 3]
```diff
@@ -319,7 +319,7 @@ def _tie_or_clone_weights(self, output_embeddings, input_embeddings):
         if hasattr(output_embeddings, "out_features") and hasattr(input_embeddings, "num_embeddings"):
             output_embeddings.out_features = input_embeddings.num_embeddings
 
-    def resize_token_embeddings(self, new_num_tokens=None):
+    def resize_token_embeddings(self, new_num_tokens: int = None):
```
Comment (julien-c) [inline src/transformers/modeling_utils.py:322]: `Optional[int] = None`?

File: src/transformers/modeling_utils.py [hunk 4]
```diff
@@ -398,7 +402,7 @@ def init_weights(self):
         # Tie weights if needed
         self.tie_weights()
 
-    def prune_heads(self, heads_to_prune):
+    def prune_heads(self, heads_to_prune: dict):
```
Comment (julien-c) [inline src/transformers/modeling_utils.py:405]:
```suggestion
    def prune_heads(self, heads_to_prune: Dict):
```

File: src/transformers/modeling_utils.py [hunk 5]
```diff
@@ -763,27 +767,27 @@ def enforce_repetition_penalty_(self, lprobs, batch_size, num_beams, prev_output
...
+        bad_words_ids: Sequence = None,
+        bos_token_id: int = None,
+        pad_token_id: int = None,
+        eos_token_id: int = None,
+        length_penalty: float = None,
+        no_repeat_ngram_size: int = None,
+        num_return_sequences: int = None,
+        attention_mask: torch.LongTensor = None,
+        decoder_start_token_id: int = None,
+        use_cache: bool = None,
```
Comment (julien-c) [inline src/transformers/modeling_utils.py:789]: All of those should be `Optional[...]`

File: src/transformers/modeling_utils.py [hunk 6]
```diff
@@ -1563,7 +1567,7 @@ def _get_generated_ngrams(hypo_idx):
     return banned_tokens
 
 
-def calc_banned_bad_words_ids(prev_input_ids, bad_words_ids):
+def calc_banned_bad_words_ids(prev_input_ids: Sequence, bad_words_ids: Sequence) -> Sequence:
```
Comment (julien-c) [inline src/transformers/modeling_utils.py:1570]: Parametrize by their generic type, i.e. `Sequence[int]`

**Review by julien-c — CHANGES_REQUESTED** [review]

File: src/transformers/modeling_utils.py (hunk 2 above)
Comment (bglearning) [reply src/transformers/modeling_utils.py:270]: Going through some tests, looks like it's expected to be either `nn.Embedding` or `nn.AdaptiveEmbedding`. The docstring has `nn.Module` (as "A module mapping vocabulary to hidden states"). Maybe it is so to have it be general? Not sure.

File: src/transformers/modeling_utils.py (hunk 6 above)
Comment (bglearning) [reply src/transformers/modeling_utils.py]: Done

File: src/transformers/modeling_utils.py (hunk 5 above)
Comment (bglearning) [reply src/transformers/modeling_utils.py]: Done

File: src/transformers/modeling_utils.py (hunk 4 above)
Comment (bglearning) [reply src/transformers/modeling_utils.py]: Changed it.

File: src/transformers/modeling_utils.py (hunk 3 above)
Comment (bglearning) [reply src/transformers/modeling_utils.py]: Done

File: src/transformers/modeling_utils.py (hunk 1 above)
Comment (bglearning) [reply src/transformers/modeling_utils.py]: Changed to `Iterable`

File: src/transformers/modeling_utils.py [hunk 7]
```diff
@@ -175,7 +175,9 @@ def get_extended_attention_mask(self, attention_mask: Tensor, input_shape: tuple
         extended_attention_mask = (1.0 - extended_attention_mask) * -10000.0
         return extended_attention_mask
 
-    def get_head_mask(self, head_mask, num_hidden_layers, is_attention_chunked=False):
+    def get_head_mask(
+        self, head_mask: Tensor, num_hidden_layers: int, is_attention_chunked: Optional[bool] = False
+    ) -> Tensor:
```
Comment (julien-c) [inline src/transformers/modeling_utils.py:180]:
```suggestion
    def get_head_mask(self, head_mask: Tensor, num_hidden_layers: int, is_attention_chunked: bool = False) -> Tensor:
```

File: src/transformers/modeling_utils.py [hunk 8]
```diff
@@ -131,7 +131,7 @@ def invert_attention_mask(self, encoder_attention_mask: Tensor) -> Tensor:
         encoder_extended_attention_mask = (1.0 - encoder_extended_attention_mask) * -1e9
         return encoder_extended_attention_mask
 
-    def get_extended_attention_mask(self, attention_mask: Tensor, input_shape: tuple, device: device):
+    def get_extended_attention_mask(self, attention_mask: Tensor, input_shape: tuple, device: device) -> Tensor:
```
Comment (julien-c) [inline src/transformers/modeling_utils.py:134]:
```suggestion
    def get_extended_attention_mask(self, attention_mask: Tensor, input_shape: Tuple, device: device) -> Tensor:
```

**Comment by julien-c** [comment]
Thanks @bglearning!

**Pull request merged**